ADAL/src/workplacejoin/mac/ADWorkPlaceJoinUtil.m

```diff
@@ -22,14 +22,292 @@
+{
+    OSStatus status = noErr;
+    ADAuthenticationError* adError = nil;
+    NSData* issuer = nil;
+    NSMutableDictionary *identityQuery = nil;
+    CFDictionaryRef result = NULL;
+    
+    *identity = nil;
+    *clientCertificate = nil;
+    if (error)
+        *error = nil;
+
+    *clientCertificate = [self copyWPJCertificateRef:correlationId error:&adError];
+    
+    if (adError)
+    {
+        if (error)
+            *error = adError;
+        
+        AD_LOG_ERROR_F(@"Failed to retrieve WPJ client certificate from keychain - ", adError.code, nil, @"%@ correlation id.", correlationId);
+        goto _error;
+    }
+    
+    // If there's no certificate in the keychain, adError won't be set since this isn't an error condition.
+    if (!*clientCertificate)
+        return;
+    
+    // In OS X the shared access group cannot be set, so the search needs to be more
+    // specific. The code below searches the identity by passing the WPJ cert as reference.
+    identityQuery = [[NSMutableDictionary alloc] init];
+    SAFE_ARC_AUTORELEASE(identityQuery);
+    
+    [identityQuery setObject:(__bridge id)kSecClassIdentity forKey:(__bridge id)kSecClass];
+    [identityQuery setObject:(__bridge id)kCFBooleanTrue forKey:(__bridge id<NSCopying>)(kSecReturnRef)];
+    [identityQuery setObject:(__bridge id)kCFBooleanTrue forKey:(__bridge id<NSCopying>)(kSecReturnAttributes)];
+    [identityQuery setObject:(__bridge id)kSecAttrKeyClassPrivate forKey:(__bridge id)kSecAttrKeyClass];
+    [identityQuery setObject:(__bridge id)*clientCertificate forKey:(__bridge id)kSecValueRef];
+    
+    status = SecItemCopyMatching((__bridge CFDictionaryRef)identityQuery, (CFTypeRef*)&result);
+    CHECK_KEYCHAIN_STATUS(@"Failed to retrieve WPJ identity from keychain.");
+    
+    issuer = [(__bridge NSDictionary*)result objectForKey:(__bridge id)kSecAttrIssuer];
+    if (issuer)
+    {
+        *clientCertificateIssuer = [[NSString alloc] initWithData:issuer encoding:NSISOLatin1StringEncoding];
+        SAFE_ARC_AUTORELEASE(*clientCertificateIssuer);
+    }
+    
+    *identity = (__bridge SecIdentityRef)([(__bridge NSDictionary*)result objectForKey:(__bridge id)kSecValueRef]);
+    if (*identity)
+        CFRetain(*identity);
+    
+    CFRelease(result);
+    
+    return;
+    
+_error:
+    
+    if (*identity)
+        CFRelease(*identity);
+    *identity = nil;
+    
+    if (*clientCertificate)
+        CFRelease(*clientCertificate);
+    *clientCertificate = nil;
+    
+    *clientCertificateIssuer = nil;
+}
+
+
++ (SecCertificateRef) copyWPJCertificateRef:(NSUUID*)correlationId
+                                      error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    OSStatus status= noErr;
+    SecCertificateRef certRef = NULL;
+    NSData *issuerTag = [self wpjCertIssuerTag];
+    
+    NSMutableDictionary *queryCert = [[NSMutableDictionary alloc] init];
```

Comment:
`kSecAttrLabel` not being human readable on macOS is a bug either way.

As for the rest, the UPN being stored separately from the certificate is one of the more common causes of bugs on iOS and a hurdle if we ever have to support multiple WPJs. Seeing how we're not starting with any pre-existing state here we should avoid repeating the same mistakes.